### PR TITLE
[Break BC] Create training directory, move checkpointing

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
The big kahuna of refactors. 

Motivation is discussed extensively in #1421. Here, we only move the checkpointing directory into training. This should break most cycles since it depends on models. Unfortunately, this touches all configs... will run a few to make sure none break.

All references to `torchtune.utils.FullModelXXXCheckpointer` now becomes `torchtune.training.FullModelXXXCheckpointer`

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: docs/source/deep_dives/checkpointer.rst [hunk 1]
```diff
@@ -381,7 +381,7 @@ looks something like this:
     checkpointer:
 
         # checkpointer to use
-        _component_: torchtune.utils.FullModelHFCheckpointer
```
Comment (ebsmothers) [inline docs/source/deep_dives/checkpointer.rst:384]: L430 too

File: torchtune/training/checkpointing/_checkpointer.py [hunk 2]
```diff
@@ -13,13 +13,13 @@
 
 import torch
 from safetensors.torch import save_file
-from torchtune import utils
+from torchtune import training
```
Comment (ebsmothers) [inline torchtune/training/checkpointing/_checkpointer.py:16]: Not a huge deal (given the scope of other stuff you're doing here) but it's a little weird to me that we use `training.MODEL_KEY` etc. when all these things are literally defined in the local directory. Feels like needless indirection to me

**Review by ebsmothers — APPROVED** [review]
A few small comments, but modulo resolving merge conflicts and (still) green CI this looks good to me

File: torchtune/training/__init__.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    "ADAPTER_KEY",
+    "EPOCHS_KEY",
+    "MAX_STEPS_KEY",
+    "MODEL_KEY",
+    "OPT_KEY",
+    "RNG_KEY",
+    "SEED_KEY",
+    "STEPS_KEY",
+    "TOTAL_EPOCHS_KEY",
+]
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/training/__init__.py:43]: mainly to keep it consistent, but not opposed to removing these

File: docs/source/deep_dives/checkpointer.rst (hunk 1 above)
Comment (RdoubleA) [reply docs/source/deep_dives/checkpointer.rst:384]: dam good eye

File: torchtune/training/checkpointing/_checkpointer.py (hunk 2 above)
Comment (RdoubleA) [reply torchtune/training/checkpointing/_checkpointer.py:16]: eh, later problem
